### PR TITLE
#150 add save notification tab content to our system

### PR DESCRIPTION
**Pull request by SlakvichAleksandr**

### CODE REVIEW

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 1]
```diff
@@ -88,6 +90,17 @@ public async Task CreateAvailabilitySlot(SaveAvailabilitySlotDto slotDto)
...
 
+            if (slotDto.EmailSettings is not null)
+            {
+                var emailTemplate = _mapper.Map<EmailTemplate>(slotDto.EmailSettings,
+                    opts => opts.AfterMap((_, dest) =>
+                    {
+                        dest.AvailabilitySlot = entity;
+                    }));
+                await _context.EmailTemplates.AddAsync(emailTemplate);
+                entity.EmailTemplates.Add(emailTemplate);
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: Please move this to a separate method

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 2]
```diff
@@ -144,6 +159,18 @@ public async Task<AvailabilitySlotDto> UpdateAvailabilitySlot(long id, SaveAvail
                 _context.Remove(availabilitySlot.AdvancedSlotSettings);
             }
 
+            if (updateAvailabilityDto.EmailSettings is not null)
+            {
+                var newTemplate = _mapper.Map<EmailTemplate>(updateAvailabilityDto.EmailSettings);
+                var oldTemplate = await _context.EmailTemplates.Where(t => t.TemplateType == newTemplate.TemplateType && t.AvailabilitySlotId == availabilitySlot.Id).FirstOrDefaultAsync();
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]:
```suggestion
                var oldTemplate = await _context.EmailTemplates
                    .Where(t => t.TemplateType == newTemplate.TemplateType && t.AvailabilitySlotId == availabilitySlot.Id)
                    .FirstOrDefaultAsync();
```

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 3]
```diff
@@ -144,6 +159,18 @@ public async Task<AvailabilitySlotDto> UpdateAvailabilitySlot(long id, SaveAvail
...
+                if (oldTemplate != null)
+                {
+                    _context.EmailTemplates.Remove(oldTemplate);
+                }
+
+                availabilitySlot.EmailTemplates.Add(newTemplate);
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: You need to update template, not delete and add

File: backend/EasyMeets.Core/EasyMeets.Core.DAL/Context/ModelBuilderExtensions.cs [hunk 4]
```diff
@@ -21,6 +20,7 @@ public static void Configure(this ModelBuilder modelBuilder)
             modelBuilder.ApplyConfigurationsFromAssembly(typeof(ExternalAttendeeConfig).Assembly);
             modelBuilder.ApplyConfigurationsFromAssembly(typeof(MeetingConfig).Assembly);
             modelBuilder.ApplyConfigurationsFromAssembly(typeof(QuestionsConfig).Assembly);
+            modelBuilder.ApplyConfigurationsFromAssembly(typeof(EmailTemplateConfig).Assembly);
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.DAL/Context/ModelBuilderExtensions.cs]: Please delete this

File: backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Controllers/AvailabilityController.cs [hunk 5]
```diff
@@ -6,7 +6,6 @@
 
 namespace EasyMeets.Core.WebAPI.Controllers
 {
-    [Authorize]
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Controllers/AvailabilityController.cs]: Why did you delete this?

File: frontend/src/app/core/models/IAvailabilitySlot.ts [hunk 6]
```diff
@@ -26,5 +26,5 @@ export interface IAvailabilitySlot {
     passwordProtectionIsUsed: boolean;
     passwordProtection?: string;
     timeZoneVisibility: boolean;
-    confirmationEmailSettings?: ISaveConfirmationEmailDetails;
+    confirmationEmailSettings?: ISaveConfirmationEmailDetails[];
```
Comment (tatianahutii) [inline frontend/src/app/core/models/IAvailabilitySlot.ts]: It would be better to rename this field to more common

File: frontend/src/app/core/models/save-availability-slot/ISaveAvailability.ts [hunk 7]
```diff
@@ -3,11 +3,14 @@ import { ISaveEventDetails } from '@core/models/save-availability-slot/ISaveEven
...
+import { ISaveConfirmationEmailDetails } from './ISaveConfirmationEmailDetails';
+
 export interface ISaveAvailability {
     teamId?: number;
     hasAdvancedSettings: boolean;
     generalDetails: ISaveGeneralSettings;
     eventDetails: ISaveEventDetails;
     advancedSettings: ISaveAdvancedSettings | null;
     schedule: ISchedule;
+    emailSettings: ISaveConfirmationEmailDetails;
```
Comment (tatianahutii) [inline frontend/src/app/core/models/save-availability-slot/ISaveAvailability.ts]: See the previous comment

File: frontend/src/app/modules/availability/edit-slot/edit-availability-page/edit-availability-page.component.ts [hunk 8]
```diff
@@ -28,14 +28,13 @@ export class EditAvailabilityPageComponent extends BaseComponent {
...
             this.spinnerService.show();
-            this.http.getSlotById(this.id)
-                .subscribe(slotResponse => {
-                    this.slot = slotResponse;
-                    this.spinnerService.hide();
-                });
+            this.http.getSlotById(this.id).subscribe((slotResponse) => {
+                this.slot = slotResponse;
+                this.spinnerService.hide();
+            });
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/edit-slot/edit-availability-page/edit-availability-page.component.ts]: You need to unsubscribe from every subscribe

File: frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts [hunk 9]
```diff
@@ -1,41 +1,91 @@
...
+    private _confirmationSubject = 'Confirmation Email';
+
+    private _cancellationSubject = 'Cancellation Email';
+
+    private _remindersSubject = 'Reminders Email';
+
+    private _followUpSubject = 'Follow-Up Email';
+
+    private _body =
+        // eslint-disable-next-line max-len
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts]: Please do not use eslint-disable-next-line

File: frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts [hunk 10]
```diff
@@ -1,41 +1,91 @@
...
+        if (path === 'follow-up') {
+            this.activeTab = path;
+            const followUpSetting = this.slot?.confirmationEmailSettings?.find(
+                (obj) => obj.type === TemplateType.FollowUp,
+            );
+
+            this.settings.emailSubject = followUpSetting?.emailSubject ?? this._followUpSubject;
+            this.settings.type = TemplateType.FollowUp;
+            this.settings.emailBody = followUpSetting?.emailBody ?? this._body;
+        }
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts]: Please refactor these lines

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs (hunk 3 above)
Comment (SlakvichAleksandr) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: We need to have only 1 template with type(Confirmation, Cancellation, etc.) for settings. For this reason previous email template is removing

File: backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Controllers/AvailabilityController.cs (hunk 5 above)
Comment (SlakvichAleksandr) [reply backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Controllers/AvailabilityController.cs]: My bad ) for local tests )

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs (hunk 3 above)
Comment (SlakvichAleksandr) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: I will rewrite code for update previous template

File: backend/EasyMeets.Core/EasyMeets.Core.DAL/Context/ModelBuilderExtensions.cs (hunk 4 above)
Comment (SlakvichAleksandr) [reply backend/EasyMeets.Core/EasyMeets.Core.DAL/Context/ModelBuilderExtensions.cs]: Why we should delete config for EmailTemplate entity ?

Comment (tatianahutii) [reply backend/EasyMeets.Core/EasyMeets.Core.DAL/Context/ModelBuilderExtensions.cs]: Because it is a reducent

File: frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts [hunk 11]
```diff
@@ -51,41 +49,28 @@ export class NotificationEmailsComponent {
...
-            this.settings.emailBody = reminderSetting?.emailBody ?? this._body;
         }
         if (path === 'follow-up') {
-            this.activeTab = path;
-            const followUpSetting = this.slot?.confirmationEmailSettings?.find(
-                (obj) => obj.type === TemplateType.FollowUp,
-            );
-
-            this.settings.emailSubject = followUpSetting?.emailSubject ?? this._followUpSubject;
+            this.currentSetting = this.slot?.emailTemplateSettings?.find((obj) => obj.type === TemplateType.FollowUp);
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts]: I believe we can use switch/case for this and move this to a separate method

File: frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts [hunk 12]
```diff
@@ -11,10 +11,9 @@ import { ISaveConfirmationEmailDetails } from '@core/models/save-availability-sl
 export class NotificationEmailsComponent {
     @Input() set newSlot(value: IAvailabilitySlot | undefined) {
         this.slot = value;
-        this.confirmationSetting = this.slot?.confirmationEmailSettings?.find(
-            (obj) => obj.type === TemplateType.Confirmation,
-        );
-        this.settings = this.confirmationSetting ?? {
+        console.log(value);
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts]: Please do not use console.log

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/.eslintrc [hunk 13]
```diff
@@ -154,7 +154,7 @@
                         "ignoreRegExpLiterals": true
                     }
                 ],
-                "indent": ["error", 4]
+                "indent": ["error", 4, { "SwitchCase": 1 }]
```
Comment (tatianahutii) [inline frontend/.eslintrc:157]: Do not change elsint

File: frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts [hunk 14]
```diff
@@ -48,27 +47,34 @@ export class NotificationEmailsComponent {
...
+                this.settings.type = TemplateType.Reminders;
+                break;
+            case 'follow-up':
+                this.currentSetting = this.slot?.emailTemplateSettings?.find(
+                    (obj) => obj.type === TemplateType.Reminders,
+                );
+                this.settings.type = TemplateType.Reminders;
+                break;
+            default:
+                break;
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts]: Please refactor this

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/.eslintrc (hunk 13 above)
Comment (SlakvichAleksandr) [reply frontend/.eslintrc:157]: How I can resolve this eslint-prettier stupid error ?
https://prnt.sc/uYvf_dTCX0Xu

File: frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts [hunk 15]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable indent */
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts:1]: Do not use eslint-disable indent

File: frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts [hunk 16]
```diff
@@ -49,33 +50,23 @@ export class NotificationEmailsComponent {
...
                 this.settings.type = TemplateType.Reminders;
                 break;
             default:
                 break;
         }
         this.activeTab = path;
+        this.currentSetting = this.slot?.emailTemplateSettings?.find((obj) => obj.type === this.settings.type);
+        this.settings.allowToSend = this.currentSetting?.allowToSend ?? false;
         this.settings.emailSubject = this.currentSetting?.emailSubject ?? this._confirmationSubject;
         this.settings.emailBody = this.currentSetting?.emailBody ?? this._body;
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts]: Please get rid of mutation

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts (hunk 15 above)
Comment (tatianahutii) [reply frontend/src/app/modules/availability/new-slot/notification-emails/notification-emails.component.ts:1]: Okay, you can return indent: ["error", 2, { "SwitchCase": 1 }]
It would be better to let us know about errors and eslint config changing before operating PR

**Review by tatianahutii — APPROVED** [review]
